server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -688,6 +748,20 @@ public void finalizeSnapshot(final SnapshotId snapshotId,
         }));
     }
 
+    // Delete all old shard gen blobs that aren't referenced any longer as a result from moving to updated repository data
+    private void cleanupOldShardGens(RepositoryData existingRepositoryData, RepositoryData updatedRepositoryData) {
+        final List<String> toDelete = new ArrayList<>();
+        final int prefixPathLen = basePath().buildAsString().length();
+        updatedRepositoryData.shardGenerations().obsoleteShardGenerations(existingRepositoryData.shardGenerations()).forEach(
+            (indexId, gens) -> gens.forEach((shardId, oldGen) -> toDelete.add(
+                shardContainer(indexId, shardId).path().buildAsString().substring(prefixPathLen) + INDEX_FILE_PREFIX + oldGen)));
```

Comment:
Sure handle that in a follow up (wanted to do that anyway since exposing the real path is a bit risky).

We do have this tested properly  I think. It passes all 3rd party tests (which all include cleanup tests that make sure deletes actually worked out).

Reply:
Should we extend the repo consistency check so that all the shard generations mentioned in the repository data exist?

Reply:
Yea def. we should. I'll add that.

Reply:
Ok I addressed this now in 1dffe60730d together with addressing the issue of reducing shard count by introducing special shard generations for deleted and known-to-not-exist shards.
Hope the added documentation makes things clear.

Also, I found a "bug" in the existing implementation that unfortunately I didn't find a clean way to solve here either. See the TODO in that commit, we currently fail to delete empty shard directories (just the directory itself not the contents). I didn't fix that here since it's a very minor issue, only affects FS+HDFS repos and is very tricky to fix cleanly.